### PR TITLE
Add BPoly to cupyx.scipy.interpolate module

### PR DESCRIPTION
**Pull request by andfoy**
See https://github.com/cupy/cupy/issues/7186

### CODE REVIEW

File: cupyx/scipy/interpolate/_interpolate.py [hunk 1]
```diff
@@ -294,6 +295,132 @@
...
+        res = coef[i0] * s1 * s1 + coef[i1] * 2.0 * s1 * s + coef[i2] * s * s;
+    } else if(k == 3) {
+        res = (coef[i0] * s1 * s1 * s1 + coef[i1] * 3.0 * s1 * s1 * s +
+               coef[i2] * 3.0 * s1 * s * s + coef[i3] * s * s * s);
+    } else {
+        T comb = 1;
+        res = 0;
+        for(int j = 0; j < k + 1; j++) {
+            const long long idx = j * c_strides[0] + ci * c_strides[1] + cj;
+            res += comb * pow(s, j) * pow(s1, ((int) k) - j) * coef[idx];
```
Comment (andfoy) [inline cupyx/scipy/interpolate/_interpolate.py:330]: @takagi, do you know if `pow` is valid here? If I try to compile the kernel, then the compiler will complain about not finding an adequate symbol definition for `pow`:

```python
  File "/home/andfoy/Documentos/Quansight/cupy/cupy/cuda/compiler.py", line 322, in compile_using_nvrtc
    return _compile(source, options, cu_path,
  File "/home/andfoy/Documentos/Quansight/cupy/cupy/cuda/compiler.py", line 306, in _compile
    compiled_obj, mapping = prog.compile(options, log_stream)
  File "/home/andfoy/Documentos/Quansight/cupy/cupy/cuda/compiler.py", line 705, in compile
    raise CompileException(log, self.src, self.name, options,
cupy.cuda.compiler.CompileException: ptxas fatal   : Unresolved extern function '_Z4powidi'
```

If I try to use either `-dc` or `-rdc=true`, then the kernel will be marked as invalid:

```python
File "cupy/_core/core.pyx", line 2265, in cupy._core.core.compile_with_cache
    return cuda.compiler._compile_module_with_cache(
  File "/home/andfoy/Documentos/Quansight/cupy/cupy/cuda/compiler.py", line 493, in _compile_module_with_cache
    return _compile_with_cache_cuda(
  File "/home/andfoy/Documentos/Quansight/cupy/cupy/cuda/compiler.py", line 580, in _compile_with_cache_cuda
    cubin = ls.complete()
  File "cupy/cuda/function.pyx", line 306, in cupy.cuda.function.LinkState.complete
    cpdef bytes complete(self):
  File "cupy/cuda/function.pyx", line 307, in cupy.cuda.function.LinkState.complete
    cubin = driver.linkComplete(self.ptr)
  File "cupy_backends/cuda/api/driver.pyx", line 185, in cupy_backends.cuda.api.driver.linkComplete
    check_status(status)
  File "cupy_backends/cuda/api/driver.pyx", line 60, in cupy_backends.cuda.api.driver.check_status
    raise CUDADriverError(status)
cupy_backends.cuda.api.driver.CUDADriverError: CUDA_ERROR_INVALID_SOURCE: device kernel image is invalid
```

Comment (kmaehashi) [reply cupyx/scipy/interpolate/_interpolate.py:330]: `pow` only accepts `(double, double)` in CUDA:
https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__DOUBLE.html#group__CUDA__MATH__DOUBLE_1g6d36757715384dc18e0483aa1f04f6c7

c.f. https://github.com/cupy/cupy/blob/46f632218f7ccedffd7182eb37294229e25e3aee/cupy/_core/_routines_math.pyx#L1007-L1022

(thanks @emcastillo for the hint 😃)

Comment (andfoy) [reply cupyx/scipy/interpolate/_interpolate.py:330]: Thanks a lot for the pointers!

**Comment by andfoy** [comment]
These are the benchmark results that compare the current `BPoly` implementation (as of [00a3367](https://github.com/cupy/cupy/pull/7343/commits/00a3367fd6cb573b6950c9b1d4c588e7e4220ab2)) against the SciPy one. For larger inputs, CuPy implementation presents a ~13X improvement over the Scipy one.

<details>
<summary>Benchmark script</summary>

```python
from itertools import product

import cupy as cp
import numpy as np
from cupy import testing
from cupyx.profiler import benchmark

from cupyx.scipy.interpolate import BPoly as CuBPoly
from scipy.interpolate import BPoly as CPUBPoly

from tqdm import tqdm


input_sizes = [5, 10, 100, 1000, 5000, 10000, 50000, 100000, 500000, 700000]
modules = [(cp, CuBPoly, 'CuPy'), (np, CPUBPoly, 'SciPy')]
kwargs = [('extrapolate', [True, False, 'periodic'])]
# kwargs = [('extrapolate', [True])]
kwargs = list(product(*[list(product([x[0]], x[1])) for x in kwargs]))
measurement = {}


def gather_time(prof):
    cpu_time = prof.cpu_times.mean() * 1000
    gpu_time = prof.gpu_times.mean() * 1000
    return max(cpu_time, gpu_time)


def input_creation(xp, cls, size, **kwargs):
    order = 3
    rand_sample = testing.shaped_random((30,), xp, scale=1)
    x = xp.unique(xp.r_[0, 10 * rand_sample, 10])
    c = testing.shaped_random((order + 1, len(x) - 1, 2, 3), xp, scale=2) - 1
    b = cls(c, x, **kwargs)
    x = testing.shaped_random(size, xp)
    return b, x


def closure(b, x):
    return b(x)


for kwarg_comb in kwargs:
    kwarg_measurement = measurement.get(kwarg_comb, {})
    measurement[kwarg_comb] = kwarg_measurement
    kwarg_comb = dict(kwarg_comb)
    for size in tqdm(input_sizes):
        size_measurement = kwarg_measurement.get(size, {})
        kwarg_measurement[size] = size_measurement
        for xp, cls, _id in modules:
            b, x = input_creation(xp, cls, (size,), **kwarg_comb)
            prof = benchmark(closure, (b, x), n_repeat=100)
            size_measurement[_id] = gather_time(prof)


lines = []
for kwarg_comb in kwargs:
    kwarg_line = ', '.join([f'{x}={y}' for x, y in kwarg_comb])
    lines.append(f'### `{kwarg_line}`\n')
    lines.append('| Size | CuPy (ms) | SciPy (ms) | Speedup |')
    lines.append('|:----:|:---------:|:----------:|:-------:|')
    kwarg_measurement = measurement[kwarg_comb]
    for size in input_sizes:
        comp = f'{size}'
        size_measurement = kwarg_measurement[size]
        times = []
        for _, _, _id in modules:
            time = size_measurement[_id]
            times.append(time)
            time_info = f'{time:3f}'
            comp = f'{comp} | {time_info}'
        speedup = times[1] / times[0]
        comp = f'{comp} | {speedup:3f}'
        lines.append(f'| {comp} |')
    lines.append('\n')


print('\n'.join(lines))
```
</details>

### `extrapolate=True`

| Size | CuPy (ms) | SciPy (ms) | Speedup |
|:----:|:---------:|:----------:|:-------:|
| 5 | 0.431522 | 0.011833 | 0.027422 |
| 10 | 0.458396 | 0.012253 | 0.026731 |
| 100 | 0.444468 | 0.019361 | 0.043560 |
| 1000 | 0.451679 | 0.099132 | 0.219473 |
| 5000 | 0.466499 | 0.418778 | 0.897703 |
| 10000 | 0.514298 | 0.824277 | 1.602725 |
| 50000 | 0.837968 | 4.077825 | 4.866324 |
| 100000 | 1.262015 | 8.136389 | 6.447143 |
| 500000 | 3.550142 | 40.689160 | 11.461276 |
| 700000 | 4.164489 | 59.602877 | 14.312171 |


### `extrapolate=False`

| Size | CuPy (ms) | SciPy (ms) | Speedup |
|:----:|:---------:|:----------:|:-------:|
| 5 | 0.505202 | 0.011811 | 0.023379 |
| 10 | 0.428217 | 0.012354 | 0.028849 |
| 100 | 0.437529 | 0.039578 | 0.090459 |
| 1000 | 0.454169 | 0.091461 | 0.201382 |
| 5000 | 0.494044 | 0.425856 | 0.861981 |
| 10000 | 0.500459 | 0.823594 | 1.645676 |
| 50000 | 0.854928 | 4.088470 | 4.782239 |
| 100000 | 1.241419 | 8.142378 | 6.558929 |
| 500000 | 3.698917 | 40.683270 | 10.998697 |
| 700000 | 4.282309 | 59.551235 | 13.906337 |


### `extrapolate=periodic`

| Size | CuPy (ms) | SciPy (ms) | Speedup |
|:----:|:---------:|:----------:|:-------:|
| 5 | 0.505506 | 0.019109 | 0.037803 |
| 10 | 0.529870 | 0.019564 | 0.036921 |
| 100 | 0.503514 | 0.027705 | 0.055023 |
| 1000 | 0.541925 | 0.106474 | 0.196473 |
| 5000 | 0.543730 | 0.455507 | 0.837744 |
| 10000 | 0.615191 | 0.903536 | 1.468709 |
| 50000 | 0.915069 | 4.370491 | 4.776130 |
| 100000 | 1.336796 | 8.737269 | 6.535978 |
| 500000 | 3.137338 | 43.475502 | 13.857449 |
| 700000 | 4.564795 | 63.811410 | 13.979030 |

**Comment by takagi** [comment]
Thanks for the benchmark, It's very nice!

File: cupyx/scipy/interpolate/_interpolate.py [hunk 2]
```diff
@@ -976,3 +1172,592 @@ def from_spline(cls, tck, extrapolate=None):
...
+        """
+        Construct a piecewise polynomial in the Bernstein basis,
+        compatible with the specified values and derivatives at breakpoints.
+
+        Parameters
+        ----------
+        xi : array_like
+            sorted 1-D array of x-coordinates
+        yi : array_like or list of array_likes
+            ``yi[i][j]`` is the ``j``th derivative known at ``xi[i]``
```
Comment (takagi) [inline cupyx/scipy/interpolate/_interpolate.py:1569]: maybe a space is needed after ``. (SciPy's also broken)
```suggestion
            ``yi[i][j]`` is the ``j`` th derivative known at ``xi[i]``
```

File: cupyx/scipy/interpolate/_interpolate.py [hunk 3]
```diff
@@ -976,3 +1172,592 @@ def from_spline(cls, tck, extrapolate=None):
...
+        This uses the fact that a Bernstein polynomial `b_{a, k}` can be
+        identically represented as a linear combination of polynomials of
+        a higher degree `k+d`:
+
+            .. math:: b_{a, k} = comb(k, a) \sum_{j=0}^{d} b_{a+j, k+d} \
+                                 comb(d, j) / comb(k+d, a+j)
+        """
+        if comb is None:
+            raise NotImplementedError(
+                'raise_degree requires a SciPy installation')
```
Comment (takagi) [inline cupyx/scipy/interpolate/_interpolate.py:1509]: Would you use `math.comb` (Python >=3.8) or a naive one (Python == 3.7) instead so we don't need to depend on scipy here?
https://github.com/cupy/cupy/blob/aa12407137ae15ad014e14842c5e2a18338716cd/cupyx/scipy/interpolate/_rbfinterp.py#L397-L402

**Review by takagi — APPROVED** [review]
LGTM!

**Review by takagi — APPROVED** [review]

**Comment by takagi** [comment]
Thanks, @andfoy!

**Pull request merged**